x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/BasicQueryClient.java

```diff
@@ -64,48 +68,52 @@ protected void search(SearchRequest search, ActionListener<SearchResponse> liste
+    public void fetchHits(Iterable<List<HitReference>> refs, ActionListener<List<List<SearchHit>>> listener) {
+        IdsQueryBuilder idsQuery = idsQuery();
 
-        int sz = 0;
+        int innerListSize = 0;
+        Set<String> indices = new HashSet<>();
+
+        // associate each reference with its own
+        final Map<HitReference, List<Integer>> referenceToPosition = new HashMap<>();
+        int counter = 0;
 
         for (List<HitReference> list : refs) {
-            sz = list.size();
+            innerListSize = list.size();
             for (HitReference ref : list) {
-                Item item = new Item(ref.index(), ref.id());
-                // make sure to get the whole source
-                item.fetchSourceContext(FetchSourceContext.FETCH_SOURCE);
-                requestBuilder.add(item);
+                idsQuery.addIds(ref.id());
+                indices.add(ref.index());
+                // remember the reference position
+                List<Integer> positions = referenceToPosition.computeIfAbsent(ref, v -> new ArrayList<>(1));
+                positions.add(counter++);
             }
         }
 
-        final int listSize = sz;
-        client.multiGet(requestBuilder.request(), wrap(r -> {
-            List<List<GetResponse>> hits = new ArrayList<>(r.getResponses().length / listSize);
-
-            List<GetResponse> sequence = new ArrayList<>(listSize);
-
-            int counter = 0;
-            for (MultiGetItemResponse mgr : r.getResponses()) {
-                if (mgr.isFailed()) {
-                    listener.onFailure(mgr.getFailure().getFailure());
-                    return;
-                }
+        SearchSourceBuilder builder = SearchSourceBuilder.searchSource()
+            // make sure to fetch the whole source
+            .fetchSource(FetchSourceContext.FETCH_SOURCE)
+            .trackTotalHits(false)
+            .trackScores(false)
+            .query(idsQuery);
+
+        final int listSize = innerListSize;
+        final int topListSize = counter / listSize;
+        // pre-allocate the response matrix
+        @SuppressWarnings({"rawtypes", "unchecked"})
+        List<SearchHit>[] hits = new List[topListSize];
+        for (int i = 0; i < hits.length; i++) {
+            hits[i] = Arrays.asList(new SearchHit[listSize]);
+        }
+        final List<List<SearchHit>> seq = Arrays.asList(hits);
 
-                sequence.add(mgr.getResponse());
+        SearchRequest search = prepareRequest(client, builder, false, indices.toArray(new String[0]));
 
-                if (++counter == listSize) {
-                    counter = 0;
-                    hits.add(sequence);
-                    sequence = new ArrayList<>(listSize);
-                }
+        search(search, ActionListeners.map(listener, r -> {
+            for (SearchHit hit : RuntimeUtils.searchHits(r)) {
+                List<Integer> positions = referenceToPosition.get(new HitReference(hit));
```

Comment:
Is it possible to use some other key for the HashMap to avoid the creation of HitReference objects?

Reply:
`HitReference` is that key - it's lightweight as it holds just the document index and id plus equals/hashcode. If there wasn't `HitReference`, we'd have to create it.